src/v/kafka/server/tests/produce_consume_test.cc

```diff
@@ -785,3 +785,111 @@ FIXTURE_TEST(test_basic_delete_around_batch, prod_consume_fixture) {
     BOOST_REQUIRE(!consumed_records.empty());
     BOOST_REQUIRE_EQUAL("key7", consumed_records[0].key);
 }
+
+FIXTURE_TEST(test_produce_bad_timestamps, prod_consume_fixture) {
+    /*
+     * this tests produces messages with timestamps in the future and in the
+     * past, and checks that the metric
+     * vectorized_kafka_rpc_produces_with_timestamps_out_of_bounds is correctly
+     * incremented.
+     */
+
+    wait_for_controller_leadership().get0();
+    start();
+    auto ntp = model::ntp(
+      model::ns("kafka"), test_topic, model::partition_id(0));
+
+    auto producer = tests::kafka_produce_transport(make_kafka_client().get());
+    producer.start().get();
+
+    auto bad_timestamps_metric
+      = app._kafka_server.local().probe().get_produce_bad_create_time();
+
+    BOOST_CHECK_EQUAL(0, bad_timestamps_metric);
+
+    producer
+      .produce_to_partition(
+        ntp.tp.topic,
+        ntp.tp.partition,
+        {
+          {"key0", "val0"},
+          {"key1", "val1"},
+          {"key2", "val2"},
+        },
+        model::to_timestamp(std::chrono::system_clock::now()))
+      .get();
+
+    BOOST_CHECK_EQUAL(0, bad_timestamps_metric);
+
+    // produce messages with a timestamp in the future, expect the metric to go
+    // up
+    producer
+      .produce_to_partition(
+        ntp.tp.topic,
+        ntp.tp.partition,
+        {
+          {"key0", "val0"},
+          {"key1", "val1"},
+          {"key2", "val2"},
+        },
+        model::to_timestamp(std::chrono::system_clock::now() + 2h))
+      .get();
```

Comment:
nit: It'd be nice to pull some of this into a helper to make the tests more readable and they looked like this:

```
produce_with_timestamp(now() + 2h);
BOOST_CHECK_LT(bad_timestamps_metric, current_bad_timestamp_metric());
```

Then this test would just be a handful of lines and easier to read :)

Reply:
done, also boost_test_info instead of comments to have context in case of failures